hazelcast/src/main/java/com/hazelcast/map/impl/MapEntries.java

```diff
@@ -35,31 +35,139 @@
  */
 public final class MapEntries implements IdentifiedDataSerializable {
 
+    private static interface ValueWrapper {
```

Comment:
Having these wrappers is not the way forward. 

The problem is that it will:
- increase the amount of memory used per map entry since an additional object + object reference.
- it will making walking through memory more complex due to the extra indirection
- it will put more pressure on the gc in case of an update due to additional litter.